File: spring_java_orm_sample/src/main/java/com/example/demo/api/annotation/Isbn.java
package com.example.demo.api.annotation;

import jakarta.validation.Constraint;
import jakarta.validation.Payload;
import jakarta.validation.constraints.Pattern;
import jakarta.validation.constraints.Size;

import java.lang.annotation.Documented;
import java.lang.annotation.ElementType;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.lang.annotation.Target;

@Documented
@Constraint(validatedBy = {})
@Size(min = 13, max = 13)
@Pattern(regexp = "\\d{13}")
@Target({ ElementType.FIELD, ElementType.METHOD, ElementType.PARAMETER, ElementType.ANNOTATION_TYPE })
@Retention(RetentionPolicy.RUNTIME)
public @interface Isbn {
    String message() default "ISBN must be 13 digits";

    Class<?>[] groups() default {};

    Class<? extends Payload>[] payload() default {};
}

File: codex_poting/kotlin/src/main/kotlin/com/example/demo/api/annotation/Isbn.kt
package com.example.demo.api.annotation

import jakarta.validation.Constraint
import jakarta.validation.Payload
import jakarta.validation.constraints.Pattern
import jakarta.validation.constraints.Size
import kotlin.reflect.KClass

@MustBeDocumented
@Constraint(validatedBy = [])
@Size(min = 13, max = 13)
@Pattern(regexp = "\\d{13}")
@Target(
    AnnotationTarget.FIELD,
    AnnotationTarget.FUNCTION,
    AnnotationTarget.PROPERTY_GETTER,
    AnnotationTarget.PROPERTY_SETTER,
    AnnotationTarget.VALUE_PARAMETER,
    AnnotationTarget.ANNOTATION_CLASS
)
@Retention(AnnotationRetention.RUNTIME)
annotation class Isbn(
    val message: String = "ISBN must be 13 digits",
    val groups: Array<KClass<*>> = [],
    val payload: Array<KClass<out Payload>> = []
)
